nextcord/application_command.py

```diff
@@ -1236,217 +863,757 @@ def error(self, callback: ApplicationErrorCallback) -> Callable:
+            Callback to create options from and invoke. If provided, it must be a coroutine function.
+        parent_cog: Optional[:class:`ClientCog`]
+            Class that the callback resides on. Will be passed into the callback if provided.
+
+        """
+        self.autocomplete_callback: Optional[Callable] = autocomplete_callback
+        self.autocomplete_options: Set[str] = set()
+        self.parent_cog: Optional[ClientCog] = parent_cog
+
+    def from_autocomplete_callback(self, callback: Callable) -> AutocompleteOptionMixin:
+        """Parses a callback meant to be the autocomplete function."""
+        self.autocomplete_callback = callback
+        if not asyncio.iscoroutinefunction(self.autocomplete_callback):
+            raise TypeError("Callback must be a coroutine")
+
+        skip_count = 2  # We skip the first and second args, they are always the Interaction and
+        #  the primary autocomplete value.
+        if self.parent_cog:
+            # If there's a parent cog, there should be a self. Skip it too.
+            skip_count += 1
+
+        for name, _ in signature(self.autocomplete_callback).parameters.items():
+            if skip_count:
+                skip_count -= 1
+            else:
+                self.autocomplete_options.add(name)
+
+        return self
+
+    async def invoke_autocomplete_callback(
+        self, interaction: Interaction, option_value: Any, **kwargs
+    ) -> None:
+        """|coro|
+        Invokes the autocomplete callback, injecting ``self`` if available.
+        """
+        if self.autocomplete_callback is None:
+            raise ValueError("Autocomplete hasn't been set for this function.")
+
+        if self.parent_cog:
+            return await self.autocomplete_callback(
+                self.parent_cog, interaction, option_value, **kwargs
+            )
+        else:
+            return await self.autocomplete_callback(interaction, option_value, **kwargs)
+
+
+class AutocompleteCommandMixin:
+    options: Dict[str, SlashCommandOption]
+    children: Dict[str, SlashApplicationSubcommand]
+    _state: ConnectionState
+
+    def __init__(self, parent_cog: Optional[ClientCog] = None):
+        """Contains code for providing autocomplete support, specifically for application commands.
+
+        If you are a normal user, you shouldn't be using this.
+
+        Parameters
+        ----------
+        parent_cog: Optional[:class:`ClientCog`]
+            Class that the callback resides on. Will be passed into the callback if provided.
+        """
+        self.parent_cog = parent_cog
+        self._temp_autocomplete_callbacks: Dict[str, Callable] = {}
+        # Why does this exist, and why is it "temp", you may ask? :class:`SlashCommandOption`'s are only available
+        # after the callback is fully parsed when the :class:`Client` or :class:`ClientCog` runs the from_callback
+        # method, thus we have to hold the decorated autocomplete callbacks temporarily until then.
```

Comment:
Comment above the line?
```suggestion
        # Why does this exist, and why is it "temp", you may ask? :class:`SlashCommandOption`'s are only available
        # after the callback is fully parsed when the :class:`Client` or :class:`ClientCog` runs the from_callback
        # method, thus we have to hold the decorated autocomplete callbacks temporarily until then.
        self._temp_autocomplete_callbacks: Dict[str, Callable] = {}
```